File: sdk/go/bundlebasesdk/function.go
package bundlebasesdk

import "github.com/apache/arrow-go/v18/arrow"

// ScalarFunction is the interface for implementing a custom scalar UDF.
type ScalarFunction interface {
	// Invoke applies the function to the given input arrays, returning an output array.
	Invoke(args []arrow.Array) (arrow.Array, error)
}

// AggregateFunction is the interface for implementing a custom aggregate UDF.
type AggregateFunction interface {
	// CreateState returns a new accumulator state (can be any type, stored opaquely).
	CreateState() (interface{}, error)

	// Accumulate adds data from input arrays into the accumulator state.
	Accumulate(state interface{}, args []arrow.Array) (interface{}, error)

	// Merge combines two accumulator states into one.
	Merge(stateA interface{}, stateB interface{}) (interface{}, error)

	// Evaluate produces a final scalar result from the accumulator state.
	Evaluate(state interface{}) (interface{}, error)
}

// FunctionProvider groups functions together with metadata for discovery.
type FunctionProvider interface {
	// Functions returns the available functions.
	Functions() map[string]interface{} // values are ScalarFunction or AggregateFunction

	// Metadata returns the function metadata for auto-discovery.
	Metadata() FunctionManifest
}

// FunctionManifest describes all functions in a provider for auto-detection.
type FunctionManifest struct {
	Functions []FunctionMeta `json:"functions"`
}

// FunctionMeta describes a single function for auto-detection.
type FunctionMeta struct {
	Name       string   `json:"name"`
	InputTypes []string `json:"input_types"`
	ReturnType string   `json:"return_type"`
	Kind       string   `json:"kind"` // "scalar" or "aggregate"
	Symbol     string   `json:"symbol,omitempty"`
}

File: sdk/java/src/main/java/com/bundlebase/sdk/Function.java
package com.bundlebase.sdk;

import com.fasterxml.jackson.annotation.JsonProperty;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.VectorSchemaRoot;

import java.util.List;
import java.util.Map;

/**
 * Interfaces for implementing custom Bundlebase functions.
 *
 * <p>Functions are either scalar (row-by-row) or aggregate (accumulating state).
 */
public class Function {

    private Function() {} // namespace only

    /**
     * A scalar function that transforms an input record batch into an output column.
     */
    public interface ScalarFunction {
        /**
         * Apply the function to the given input record batch, returning a single output column.
         *
         * @param input input record batch (may be null if no input data)
         * @return the result column
         */
        FieldVector invoke(VectorSchemaRoot input) throws Exception;
    }

    /**
     * An aggregate function that accumulates state across batches and produces a final result.
     *
     * @param <S> the accumulator state type
     */
    public interface AggregateFunction<S> {
        /**
         * Create a new accumulator state.
         */
        S createState() throws Exception;

        /**
         * Add data from an input record batch into the accumulator state.
         *
         * @param state the current accumulator state
         * @param input input record batch (may be null if no input data)
         * @return the updated accumulator state
         */
        S accumulate(S state, VectorSchemaRoot input) throws Exception;

        /**
         * Merge two accumulator states into one.
         *
         * @param stateA the first state (result is stored here)
         * @param stateB the second state (consumed)
         * @return the merged state
         */
        S merge(S stateA, S stateB) throws Exception;

        /**
         * Produce a final scalar result from the accumulator state.
         *
         * @param state the final accumulator state
         * @return the result value (Long, Double, String, or Boolean)
         */
        Object evaluate(S state) throws Exception;
    }

    /**
     * Groups functions together with metadata for discovery.
     */
    public interface FunctionProvider {
        /**
         * Return the available functions. Values must be {@link ScalarFunction}
         * or {@link AggregateFunction}.
         */
        Map<String, Object> functions();

        /**
         * Return function metadata for auto-discovery.
         */
        FunctionManifest metadata();
    }

    /**
     * Describes all functions in a provider for auto-detection.
     */
    public record FunctionManifest(List<FunctionMeta> functions) {}

    /**
     * Describes a single function for auto-detection.
     */
    public record FunctionMeta(
            String name,
            @JsonProperty("input_types") List<String> inputTypes,
            @JsonProperty("return_type") String returnType,
            String kind,
            String symbol
    ) {
        public FunctionMeta(String name, List<String> inputTypes, String returnType, String kind) {
            this(name, inputTypes, returnType, kind, null);
        }
    }
}
